### DISCUSSION TITLE
SubmitTurn command for AI

### DISCUSSION THREAD

**Discussion by neoedmund** [beyond-all-reason/RecoilEngine#274] [4. Feature Requests and Ideas]
When I think about a model that work all for human vs human , human vs AI , AI vs AI .
comes out an idea of `SubmitTurn` command. If all players in game, send the this  command, game server will process to the next frame. 
Human player never send this command. 
Then in all AI players game, the speed can be auto adjust to the max.

**Comment by neoedmund** [comment]
It's done https://github.com/neoedmund/spring/commit/f698a3e22be69cdeaf6ad0ba74c539365fa6f33d
Then AI have plenty time to think if not slow then FPS 30, while frame end earlier if all AI finished thinking.
I tested SimpleAI vs Barb. (BTW, Barb use less time than SimpleAI)
the speed is about x3. headless mode x6.
not that impressive but 5 ms per frame sounds reasonable.
parallel execution of AIs may have more speed, maybe useful for game like 8 vs 8.

**Reply by marcushutchings** [reply]
I believe BarB uses multi-threading already, which may be why it is using less frame time despite being more advanced than the simple AI. I think you'll find that the simulation will also be a significant limiting factor in the maximum sim speed you can attain while testing your AI.

Back to the request at hand.

While it is nice that you're thinking about these problems and how we can solve them, there are matters beyond direct AI concerns that govern the design of the engine. Here's a matter you would not likely be aware of:

In BAR we have a policy for the game speed to be determined by the median performance of the players - not the by the slowest. This means it is possible for players to fall behind as the simulation load increases. This is deliberate because one excessively slow machine should not ruin the game for everyone else. If an AI is running on a machine that is falling behind, then that AI is going to suffer as well. If the AI was to somehow be able to force everyone to slow down regardless then that has creates a means by which the policy can be defeated.

**Reply by neoedmund** [reply]
Yes. the policy is not bad, or it can be configured by different scenario. 
even in same policy , implementation can be different.
how it deal with slow client. hope not to send empty new frame. 
empty new frame is a bug seems to me. because it misinformed the client.
the right way I think is the server just slow down to speed it thinks okay.
then skipping some update() for the slow clients.

**Comment by sprunk** [comment]
What happens, right now, if you just set speed to x20 in an AI game? With human players it only goes to whatever multiplier people's machines can handle, according to the median rule mentioned above by @marcushutchings (which possibly means slowing down to less than 1.0 late-game). Doesn't this happen with AIs?
